components/director/pkg/applicationtenancy/directive.go

```diff
@@ -0,0 +1,234 @@
+	resp, err := next(ctx)
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while processing request: %s", err.Error())
+		return resp, err
+	}
+
+	tx, err := d.transact.Begin()
+	if err != nil {
+		log.C(ctx).WithError(err).Errorf("An error occurred while opening database transaction")
+		return nil, apperrors.NewInternalError("Unable to initialize database transaction: %s", err.Error())
+	}
+	defer d.transact.RollbackUnlessCommitted(ctx, tx)
+
+	ctx = persistence.SaveToContext(ctx, tx)
+
+	log.C(ctx).Debugf("Preparing tenant access creation for event: %s", eventType)
+
+	if eventType == graphql.EventTypeNewApplication {
+		tntIDFromContext, err := tenant.LoadFromContext(ctx)
+		if err != nil {
+			log.C(ctx).WithError(err).Errorf("An error occurred while loading tenant from context")
+			return nil, err
+		}
+
+		tntModel, err := d.tenantService.GetTenantByID(ctx, tntIDFromContext)
+		if err != nil {
+			log.C(ctx).WithError(err).Errorf("An error occurred while getting tenant with ID %s", tntIDFromContext)
+			return nil, errors.Wrapf(err, "while fetching tenant with id: %s", tntIDFromContext)
+		}
+
+		log.C(ctx).Debugf("Found a matching tenant in the database: %s", tntModel.ID)
+
+		if !(tntModel.Type == tenantpkg.ResourceGroup || tntModel.Type == tenantpkg.Folder || tntModel.Type == tenantpkg.Organization) {
+			log.C(ctx).Infof("Tenant type is %s. Will not continue with tanancy synchronization", tntModel.Type)
+			return resp, nil
+		}
+
+		entity, ok := resp.(graphql.Entity)
+		if !ok {
+			log.C(ctx).WithError(err).Errorf("An error occurred while casting the graphql response entity")
+			return nil, apperrors.NewInvalidDataError("An error occurred while casting the response entity: %v", resp)
+		}
+
+		if err := d.createTenantAccessForNewApplication(ctx, tntModel, entity.GetID()); err != nil {
+			return nil, err
+		}
+	} else if eventType == graphql.EventTypeNewSingleTenant {
+		tenantID, ok := resp.(string)
+		if !ok {
+			log.C(ctx).WithError(err).Errorf("An error occurred while casting the graphql response entity")
+			return nil, apperrors.NewInvalidDataError("An error occurred while casting the response entity: %v", resp)
+		}
+
+		tntModel, err := d.getTenantAndValidateTenantAccessEligibility(ctx, tenantID)
+		if err != nil {
+			return nil, err
+		}
+
+		if tntModel == nil {
+			return resp, nil
+		}
+
+		if err := d.createTenantAccessForOrgApplications(ctx, tntModel.Parent, tntModel.ID); err != nil {
+			return nil, err
+		}
+	} else if eventType == graphql.EventTypeNewMultipleTenants {
+		tenantIDs, ok := resp.([]string)
+		if !ok {
+			log.C(ctx).WithError(err).Errorf("An error occurred while casting the response entity: %v", err)
+			return nil, apperrors.NewInvalidDataError("An error occurred while casting the response entity: %v", resp)
+		}
+
+		filteredTenants := make([]*model.BusinessTenantMapping, 0)
```

Comment:
[Code duplication] Extract the code for eventType == graphql.EventTypeNewSingleTenant into a method processSingleTenantEvent(tenantID) and then
for _, tenantID := range tenantIDs{
   processSingleTenantEvent(tenantID) 
}